engine/cdn/cdn_log.go

```diff
@@ -324,3 +272,135 @@ func (s *Service) getHatchery(ctx context.Context, hatcheryID int64, hatcheryNam
 	logCache.Set(fmt.Sprintf("hatchery-key-%d", hatcheryID), pk, gocache.DefaultExpiration)
 	return pk, nil
 }
+
+func (s *Service) waitingJobs(ctx context.Context) {
+	for {
+		if ctx.Err() != nil {
+			return
+		}
+		// List all queues
+		keyListQueue := cache.Key(keyJobLogQueue, "*")
+		listKeys, err := s.Cache.Keys(keyListQueue)
+		if err != nil {
+			log.Error(ctx, "unable to list jobs queues %s", keyListQueue)
+			continue
+		}
+
+		// For each key, check if heartbeat key exist
+		for _, k := range listKeys {
+			keyParts := strings.Split(k, ":")
+			jobID := keyParts[len(keyParts)-1]
+
+			jobQueueKey, err := s.canDequeue(jobID)
+			if err != nil {
+				log.Error(ctx, "unable to check canDequeue %s: %v", jobQueueKey, err)
+				continue
+			}
+			if jobQueueKey == "" {
+				continue
+			}
+
+			sdk.GoRoutine(ctx, "cdn-dequeue-job-message", func(ctx context.Context) {
+				if err := s.dequeueJobMessages(ctx, jobQueueKey, jobID); err != nil {
+					log.Error(ctx, "unable to dequeue redis incoming job queue: %v", err)
+				}
+			})
+		}
+	}
+}
+
+func (s *Service) dequeueJobMessages(ctx context.Context, jobLogsQueueKey string, jobID string) error {
+	log.Info(ctx, "Dequeue %s", jobLogsQueueKey)
+	var t0 = time.Now()
+	var t1 = time.Now()
+	var nbMessages int
+	defer func() {
+		delta := t1.Sub(t0)
+		log.Info(ctx, "processLogs[%s] - %d messages received in %v", jobLogsQueueKey, nbMessages, delta)
+	}()
+
+	defer func() {
+		// Remove heartbeat
+		_ = s.Cache.Delete(cache.Key(keyJobHearbeat, jobID))
+	}()
+
+	tick := time.NewTicker(5 * time.Second)
+	for {
+		select {
+		case <-ctx.Done():
+			return ctx.Err()
+		case <-tick.C:
+			b, err := s.Cache.Exist(jobLogsQueueKey)
+			if err != nil {
+				log.Error(ctx, "unable to check if queue still exist: %v", err)
+				continue
+			} else if !b {
+				// leave dequeue if queue does not exist anymore
+				log.Info(ctx, "leaving job queue %s (queue no more exists)", jobLogsQueueKey)
+				return nil
+			}
+			// heartbeat
+			heartbeatKey := cache.Key(keyJobHearbeat, jobID)
+			if err := s.Cache.SetWithTTL(heartbeatKey, true, 30); err != nil {
+				log.Error(ctx, "unable to hearbeat %s: %v", heartbeatKey, err)
+				continue
+			}
+		default:
+			dequeuCtx, cancel := context.WithTimeout(ctx, 15*time.Second)
```

Comment:
What if for some reasons the heartbeat is not send or to slow, looks like another process will be able to dequeue but this one will not know about it.

Reply:
yes, he have 6 chances to heartbeat before another cdn take the queue